### PR TITLE
[WIP] Add ablecop

### PR DESCRIPTION
**Pull request by mpotter**
**TODO**
- [ ] Create install task that generates the initial config files for the application
- [ ] Add appropriate config files to .gitignore
- [x] ~~Add [pronto-rails_best_practices](https://github.com/mmozuras/pronto-rails_best_practices) ?~~
- [x] ~~Add [pronto-reek](https://github.com/mmozuras/pronto-reek) ?~~
- [x] ~~Generate pronto output as a CircleCI artifact and link from status message~~
- [ ] Readme updates
- [ ] Test coverage

### CODE REVIEW

**Comment by dennmart** [comment]
@mpotter I'm adding a couple of more things here:

* Your [pull request](https://github.com/raimondasv/pronto-rails_schema/pull/2) was merged and a new gem was released, so we're using the released gem instead of your fork.
* I added a Rake task to run the analyzers in the local branch (no need to wait for CircleCI's build).
* I added the `railties` gem as a dependency, since we need it for testing.

Currently, I ran some tests on the `master` branch of the core.able.co repo and adding this gem, and it works as it is, but I have some suggestions / questions:

* Is this going to be a public gem? The README makes it sound like it is, but to me the gem is more suited for private use. I wanted to update the README but I didn't know which direction you're planning for this gem.
* My preference with gems is to not have them generate files automatically in my projects, as it's being done here. We should have a Rake task or generator (like [RSpec](https://github.com/rspec/rspec-rails/blob/master/lib/generators/rspec/install/install_generator.rb)) that will generate the files after install. I'm not sure if that's what you meant with the first task you have in this issue.
* All the other tasks (adding other runners, generating artifacts in CircleCI) would be best to be done separately so this pull request doesn't become too huge.

File: ablecop/ablecop.gemspec [hunk 1]
```diff
@@ -30,14 +30,14 @@ Gem::Specification.new do |spec|
...
   # Fasterer will suggest some speed improvements.
   spec.add_dependency "fasterer"
   spec.add_dependency "pronto-fasterer"
 
   # SCSS Lint is a static code analyzer based on our SCSS style guide.
   spec.add_dependency "scss_lint"
   spec.add_dependency "pronto-scss"
+
+  # Pronto runner for monitoring Rails schema.rb or structure.sql consistency.
+  spec.add_dependency "pronto-rails_schema"
```
Comment (mpotter) [inline ablecop/ablecop.gemspec]: We may want to specify a version dependency. In fact, we might want to do that with all the dependencies?

**Comment by mpotter** [comment]
> Is this going to be a public gem? The README makes it sound like it is, but to me the gem is more suited for private use. I wanted to update the README but I didn't know which direction you're planning for this gem.

No intention for it to be public. This _repo_ is public for ease in deployment. I think we'd prefer it to be private until we make a conscious investment otherwise.

> My preference with gems is to not have them generate files automatically in my projects, as it's being done here. We should have a Rake task or generator (like RSpec) that will generate the files after install. I'm not sure if that's what you meant with the first task you have in this issue.

Yeah, I get that. See [my comment](https://github.com/ableco/core.able.co/pull/470#issuecomment-212916620) in Core for the rationale. You are correct in interpreting the first task, I'd want some kind of post-install generator to initially copy the files over to the project's folder and (if we stick with this pattern) add them to the `.gitignore`.

> All the other tasks (adding other runners, generating artifacts in CircleCI) would be best to be done separately so this pull request doesn't become too huge.

I think that's fine! I mostly wanted to stress test the concept of this as an independent gem. Make sure I had designed it properly and that we could accomplish the things we want to eventually accomplish. I added these tasks in as a gut-check for that but, I think it's equally cool to pursue that in subsequent PRs.

:denmark: == :hero:

File: ablecop/ablecop.gemspec [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  # Fasterer will suggest some speed improvements.
+  spec.add_dependency "fasterer"
+  spec.add_dependency "pronto-fasterer"
+
+  # SCSS Lint is a static code analyzer based on our SCSS style guide.
+  spec.add_dependency "scss_lint"
+  spec.add_dependency "pronto-scss"
+
+  # Pronto runner for monitoring Rails schema.rb or structure.sql consistency.
+  spec.add_dependency "pronto-rails_schema"
```
Comment (mpotter) [inline ablecop/ablecop.gemspec]: Should we be specifying versions in the dependencies?

**Comment by dennmart** [comment]
@mpotter As it stands now, this gem looks good to me. But I'm still not convinced that automatically creating the config files and ignoring them in the repos as you did in https://github.com/ableco/core.able.co/pull/470 is the way to go.

For example, if we have to change any of the configuration files for the runners like Rubocop, that means we need to update the gem, then update ablecop on the repos. Also, I have a feeling that even though we want to keep the code analysis rules consistent between projects, there will eventually be a case where a rule should be turned on / off for a specific project, which again won't be possible without a gem update.

I'm still vouching for having a Rake task or Rails generator to set things up and have them committed on the repo, something among the lines of:

* Add ablecop to Gemfile and do a `bundle install`.
* Run command (Rake task or Rails generator) to do initial install, which can:
  * Copy configuration files into project repo.
  * If circleci.yml exists, add the post-build task to run Pronto.
* Commit the changes to the repo.

Product engineers should be vigilant about changes made to these files if we want to keep them mostly consistent across Able's projects.

In any case, if you would prefer to keep the configuration file creation to be automatic and ignored in the repos, the gem can be merged and be used, since it works as it is.

**Comment by mpotter** [comment]
> But I'm still not convinced that automatically creating the config files and ignoring them in the repos as you did in ableco/core.able.co#470 is the way to go.

Me neither! I'd understand if we went the way you're describing, I don't think that would be a bad approach at all. Perhaps if the philosophy is such that we want global consistency, this functionality shouldn't be packaged as a gem but rather a service like [HoundCI](https://houndci.com). That would require more work so I'm not sure it's a good first step vs. a gem like this.

Maybe this would be a good debate to open up to the large team? Anyone have a POV they'd like to share? cc @ableco/engineering

**Comment by blim8183** [comment]
I like the idea of ensuring consistency across apps by not allowing configuration of the gem... I think @mpotter mentioned that there might be a desire to ignore certain files, perhaps the README could include instructions for only doing that on a per project basis?

**Comment by juliogarciag** [comment]
I'm new in this conversation, but maybe this is helpful: https://github.com/bbatsov/rubocop#inheriting-configuration-from-a-remote-url. Rubocop allows to use a remote url as parent configuration than can be useful to maintain a certain degree of consistency without manually comparison between the rubocop file of the gem and the ones in every project.

**Comment by mpotter** [comment]
@juliogarciag I saw that and I agree it's worth consideration. I opted not to go that way because, while rubocop doesn't support a remote URL, most of the other linters do not (`scss-lint`, for instance). So, I decided it was best to use a pattern that work for all of the configuration files.

**Comment by juliogarciag** [comment]
@mpotter Oh, i understand. Maybe a gem would be enough. We could use a generator to install and then propose a process to update the project with the new linter files that can come when this gem is updated. That process should take into account possible differences between what was added in the linter files of a project and the linter files in the gem, and allow somebody to make the proper decisions. I don't know if we should allow these per-project modifications on the linter files or enforce this decisions into all the projects.

File: ablecop/config/rubocop.yml [hunk 3]
```diff
@@ -0,0 +1,1136 @@
...
+    - '**/Podfile'
+    - '**/Thorfile'
+    - '**/Vagrantfile'
+    - '**/Berksfile'
+    - '**/Cheffile'
+    - '**/Vagabondfile'
+  Exclude:
+    - 'bin/**/*'
+    - 'vendor/**/*'
+    - 'db/migrate/*'
```
Comment (rezart) [inline ablecop/config/rubocop.yml]: For PICI we used a pattern of seed data going inside  `db/seeds/[environment]`. Could we add `db/seeds/*` to the exclude list?

**Comment by ccarruitero** [comment]
Well, I'm not an expert on this kind of configurations, but looks good overall for me.

I'm not used HoundCI before, but looks that a service like this could be the best option for long term.
Aside `fasterer` they miss other configuration that we plan to use?

About to keep the `repo` public or private. I'm not had a strong conviction by neither, but what about to have both kind of gems?

**Comment by dennmart** [comment]
After reading all these comments, I'm now split between two thoughts regarding the purpose of gem:

* Is the purpose of this gem to make it easy for us to set up Pronto and its runners in our Rails repos? If so, to me it shouldn't matter if people change the default configuration the gem provides to the project's specific needs. Having this as a gem would work well here, generating and keeping the configuration in the repo like I commented on yesterday.
* Is the purpose of this gem to make sure we consistently enforce the same code analysis in all our projects? If so, a gem doesn't sound like the way to go, since it seems it'll be a bit of a pain to maintain this consistency due to the different types of libraries and configurations required to be set up.

**Comment by mpotter** [comment]
@dennmart I think the issue I have is that the gem is intending to serve _both_ of the purposes you're touching on. The spec for me is:

- Quick, consistent setup for Pronto feedback on our Rails applications
- Ensure consistent conventions across our Rails applications, enforced by CircleCI + Pronto
- When a convention changes for Able, it should only need to be changed in one place, making it easy (automatic?) for that convention to be rolled out to all of our projects

The reason I went down the path of a gem:

1. It would make it trivial for developers to run the checks that would run on CI locally
2. It's easier to setup and maintain than a dedicated service

>  If so, a gem doesn't sound like the way to go, since it seems it'll be a bit of a pain to maintain this consistency due to the different types of libraries and configurations required to be set up.

I'm not sure I see the conflict as clearly as you seem to. Can you give an example of where this conflict might arise? 

Anyway, to address this potential conflict, **I'm wondering if we shouldn't consider an explicit method for overriding the gem's configuration?** I say _explicit_ because I think it should be very obvious to a developer the ways in which an application's conventions differ from Able's global conventions. My concern is if we just let the default configuration files be altered locally in the application that this becomes very difficult to track as someone is on-boarded to a project.

So, one idea to implement this would be, keep the pattern I have here that ensures the default configuration files are what exist in the gem, but the gem then looks for files like `.application.rubocop.yml` that override any config set in `.rubocop.yml`. These application specific files would not be ignored by git.

We could maybe take this concept further by _not_ copying the `.rubocop.yml` file from the gem to the application's root and instead somehow create a reference to it. I'm not sure all of the linters allow you to have a reference to a second configuration file, though. For example, rubocop allows you to [inherit from another configuration file](https://github.com/bbatsov/rubocop#inheritance) but scss-lint [doesn't seem to have support for that](https://github.com/brigade/scss-lint#configuration). This would seem to eliminate this option but maybe I'm missing something that would allow us to implement that way?

**Comment by joerodrig3** [comment]
This is geared more towards how we'd lint our front-end JS code. Recently I set up ESLint on Core([see eslint config here](https://github.com/ableco/core.able.co/blob/master/.eslintrc.js)). From my perspective, being able to modify our linter config for front-end development going from project to project may be necessary.

Here are a few situations where I see the config needing to be updated in some way:

- If we leverage another front-end library for a project, we may need to add an additional style-guide and/or plugin:
```
"extends": ["airbnb", "plugin:react/recommended"],
"plugins": ["react"],
```

- Global variables being used may change from project to project. In Core we use the google session API via Ajax calls. We declare variables attached to our window globally, so our ESLint doesn't throw an error:
```
  "globals": {
    "React": true,
    "$": true,
    "jQuery": true,
    "_": true,

    // Google session API stuff
    "gon": true,
    "gapi": true,
  },
```

- There could also be instances where depending on the project, we want to modify rules so they display a warning instead of an error, or turn off the linter entirely:

For example, in Core we use React's [dangerouslySetInnterHTML](https://github.com/ableco/core.able.co/blob/master/app/assets/javascripts/components/Utilities/CollapseHTML/CollapseHTML.js.jsx#L73) method. Since we're relying on RedCarpet to pass us sanitized HTML, we know the HTML being used is safe. By default, our React style-guide throws an error for using that property. Depending on the type of project we are working on and data we're handling, there are cases where we may want that error to be shown, whereas in other cases a warning or no notice at all may make more sense.

There are a few other cases I've seen come up by implementing the current config into Pulse and Core.

**Comment by dennmart** [comment]
@mpotter The message left by @joerodrig3 is what I was thinking about when I said it would be a pain to maintain consistency. While for the most part we'll be able to agree on the bulk of the code analysis rules for most or all our projects, there will inevitably be occasions that will need a variation of those base rules. I honestly don't think there's any way around it, unless we use the same libraries and patterns everywhere, which isn't really feasible.

I think being so strict will cause more harm than good in the long run, either by having to update the gem to accommodate the rules of a specific use case, or forcing people to use one library / pattern over another because of the enforcement of these rules.

Having an explicit way that clearly shows a deviation between the standard code rules and possible overrides is a good idea, but as you pointed out, there isn't a standard way on doing this for all the runners that we're using with Pronto. Only Rubocop allows merging of multiple configuration files, while scss-lint and fasterer don't. We could check on the existence of a configuration file in these cases in the repo, but at this point it's probably best to just have the files in the repo to begin with - my original recommendation, which I'm still preferring at this time.

**Comment by mpotter** [comment]
@dennmart Not to over-architect it but if we both agree it would be better to have explicit deviations, why not add support for that via the gem?

Our gem would write the configuration files dynamically (as it does now) using a merge of the Able defaults (maintained in the gem) and an explicit overrides file (if present) in the application folder. Given that all the linters are configured via YAML I think this would be very simple.

Pseudo-code:

```ruby
default_config = YAML.read("./config/.rubocop.yml")
application_config = YAML.read(".rubocop-application.yml")
default_config.deep_merge!(application_config)
File.write(".rubocop.yml", default_config.to_yml)
```

Thoughts?

**Comment by dennmart** [comment]
I played around with ActiveSupport's [`deep_merge`](http://api.rubyonrails.org/classes/Hash.html#method-i-deep_merge), and it worked in most cases but there were some cases that wouldn't work for us. The main scenario is that using `deep_merge`, there doesn't seem to be a straight-forward way of removing an element in the merge.

I explored a bit more and found another gem called [deep_merge](https://github.com/danielsdeleo/deep_merge) which has additional functionality over ActiveSupport's method. Particularly, it has the ability to specify elements for removal. So I added the gem and updated the `ensure_config_file!` method in the gem.

The way it will work is that it will check if the application root has a file following the pattern of `.application-<filename>` (like `.application-rubocop.yml`). If it exists, it will read the default configuration and the over-ride configuration as hashes and merge them. It will them write the merged hash into the application's configuration file in the root (if it differs from the current configuration file).

I'm still not a huge fan of the dynamic file generation, since it seems to require an application restart if we want to run that code again (like if we update the over-ride or accidentally delete one of the files). But I've complained enough about this already so I'll stop. :smirk_cat: It'll probably work okay for testing purposes.

The last bits to make a beta release for this would be adding the configuration files to the application's `.gitignore` automatically, updating the README and writing a lot of tests, which will probably take a while since I have to figure out how to test the dynamic file creation.

**Comment by blim8183** [comment]
@dennmart Your opinions are appreciated and not interpreted as complaints 👍 I think your solution sounds like a decent compromise and it's something we can evolve if we find that it's cumbersome. This gem will hopefully be an important part of our Engineering process and culture so I think we can expect that we're not handcuffed to whatever we decide now 👐 

For the files that you're proposing for over-writing configurations, is there a reason they have to be in the root? Would a folder like `config/ablecop` work better?

**Comment by mpotter** [comment]
@dennmart I don't have a POV on @blim8183's preference of putting in the root vs. otherwise but would second the sentiment. I think this might be the most substantive technical debate in Able's history... for better or worse! I'm glad to have the opportunity to debate it with you.

I also totally see your hesitation on the auto-generation of files. I agree it's an anti-pattern and one we shouldn't take lightly. I just believe that it's justified for this use case 🔥

**Comment by dennmart** [comment]
I updated the gem to look for overrides in `config/ablecop` instead of the root directory of the application. I think that's a cleaner approach.

I'm stuck and having a really difficult time doing any useful testing, due to how the files are being automatically generated. My approach was setting up a minimal dummy Rails app inside the `spec` directory, which is a common pattern for testing gems / Rails engines, and test the behavior / functionality of the gem with that. But whenever the app is loaded in the tests, the ablecop initializer will always create the files on load, which is useless since that's the behavior I want to test. I tried a few hacky things, like deleting the files and reloading the app in the tests, or waiting until the last possible moment to load the app, but the files were not being generated as they should. I'm not sure how to handle that automatic file generation in the specs.

Does anyone have suggestions on testing behavior like this, or can dig in and help with writing some?

**Comment by joerodrig3** [comment]
I'm not sure if this is in the right area, but is there any reason why we haven't adopted the safe navigation operator(released in Ruby 2.3.0) in RuboCop?

ie. 
```foo&.bar``` prevents us from having to write ```foo && foo.bar``` in a situation where foo is/can be nil

Using this results in an ```unexpected token error```

I know there is potential for it to be abused if not thought out entirely, but it is helpful in quite a few cases

**Comment by dennmart** [comment]
@joerodrig3 It seems like Rubocop doesn't currently have support for checking that (https://github.com/bbatsov/rubocop/issues/2428 has more info, and https://github.com/bbatsov/rubocop/pull/2545 seems to have been started up but never finished). When they eventually add the cop for for it, I'm all in.

**Comment by mpotter** [comment]
@dennmart Since you're having trouble with the auto-generation I put up the white flag and converted it to a generator. Will look to add specs this weekend if I find some time!

**Comment by dennmart** [comment]
@mpotter I made a couple of changes to this:

- I updated the generator methods to use methods that generators inherit from [Thor](www.rubydoc.info/github/erikhuda/thor/master/Thor/Actions.html), including updating the paths so that we don't have to rely on `Rails.root` (especially during tests).
- I included the [generator_spec](https://github.com/stevehodgkiss/generator_spec) gem to have make it easier to access to the [generator test case methods](api.rubyonrails.org/classes/Rails/Generators/TestCase.html) provided by Rails.
- Wrote specs for the generator (a hell of a lot easier this way).
- Updated the README.

I also updated the pull request on Core (https://github.com/ableco/core.able.co/pull/470) to get the latest updates as well as other README updates.

Also, since we now need to generate the files, we need to tell CircleCI to run the generator before running tests to make sure that the configuration files are set up properly before tests run. Not too happy with it but if we're going with the decision to ignore the configuration files on Git, there's no way around it.

Anything else that needs to be done here?

**Comment by mpotter** [comment]
> Also, since we now need to generate the files, we need to tell CircleCI to run the generator before running tests to make sure that the configuration files are set up properly before tests run. Not too happy with it but if we're going with the decision to ignore the configuration files on Git, there's no way around it.

I had the opportunity to debate this offline with @blim8183. The philosophy is that those files are there for the developer to use in their local IDE, as opposed to an application concern. The fact that Circle also requires them is an integration concern between the CI server and the PR review process. Again, not an application concern. Since those files should _only_ be auto-generated and not be edited, I think it's the best pattern to avoid conflicts.

> Anything else that needs to be done here?

It'd be good to get PICI as a beta user. Per Rez's [comment](https://github.com/ableco/gems.able.co/pull/3#discussion_r61153926) they need an additional ignore (which, does make sense globally). I don't have anything else!

👍

**Comment by dennmart** [comment]
That all makes sense to me, thanks for clearing that up in my head :smile_cat: 

I made a few smaller changes this morning, but nothing that changes the main functionality of the generator, so I'm going to merge this in and start using it in Core. I will then open up a pull request in the PICI repo to integrate it there for further testing.

**Pull request merged**